File: python/simulation.py
from tqdm import tqdm
import numpy as np

GRANULARITY = 0.1
GRANULARITY_INV = 10

class Time:
    def __init__(self, asint):
        self.asint = asint
        self.asfloat = asint / GRANULARITY_INV

    def __hash__(self):
        return hash(self.asint)

    def __eq__(self, other):
        return self.asint == other.asint

class Packet:
    def __init__(self, it: Time):
        self.insertion_time = it.asfloat
        self.removal_time = None

    def sent(self, t: Time):
        self.removal_time = t.asfloat
    
    def get_queueing_delay(self):
        return self.removal_time - self.insertion_time

class RouterQueue: 
    def __init__(self):
        self.queue = []
        self.queueing_delays = []

    def pop(self, t):
        if self.queue:
            packet = self.queue.pop(0)
            packet.sent(t)
            self.queueing_delays.append(packet.get_queueing_delay())
            return packet
        else:
            return None
    
    def push(self, packet):
        self.queue.append(packet)

class SingleSimulation:
    def __init__(self, A, T=1000):
        self.A = A
        self.T = T
        self.reset()

    def reset(self):
        self.queue = RouterQueue()
        self.packet_arrivals = self.generate_times()

    def simulate(self):
        num_packets_pulled = 0
        num_packets_pushed = 0
        max_queue_size = 0
        
        for t in [Time(i) for i in range(0, int(self.T * GRANULARITY_INV))]:
            for _ in range(self.packet_arrivals.get(t, 0)):
                self.queue.push(Packet(t))
                num_packets_pushed += 1

            max_queue_size = max(max_queue_size, len(self.queue.queue))

            if self.queue.pop(t) is not None:
                num_packets_pulled += 1
 
        assert num_packets_pushed == self.A * self.T
        return sum(self.queue.queueing_delays) / len(self.queue.queueing_delays), max_queue_size

    def generate_times(self):
        num_packets = self.A * self.T
        multiplier = self.T * GRANULARITY_INV

        times = (np.random.random(num_packets) * multiplier).astype(int)

        num_queued = {}
        for t in [Time(t) for t in times]:
            num_queued[t] = num_queued.get(t, 0) + 1
        
        return num_queued

class Simulation:
    def __init__(self, A, N=1000):
        self.simulations = [SingleSimulation(A) for _ in range(N)]
        self.avg_qds = []
        self.max_qs = []

    def run(self):
        for s in tqdm(self.simulations, disable=False):
            qd, qs = s.simulate()
            self.avg_qds.append(qd)
            self.max_qs.append(qs)

p25 = []
p75 = []
mean = []
qs = []
for A in range(1, 11):
    print(f"Simulating for A={A}")
    s = Simulation(A)
    s.run()
    p25.append(np.percentile(s.avg_qds, 25))
    p75.append(np.percentile(s.avg_qds, 75))
    mean.append(np.mean(s.avg_qds))
    qs.append(np.mean(s.max_qs))

print(f"p25={p25}")
print(f"p75={p75}")
print(f"mean={mean}")
print(f"qs={qs}")

File: src/simulation.rs
use rand::{Rng, distributions::Uniform};
use crate::router_queue::RouterQueue;
use std::collections::HashMap;
use crate::packet::Packet;
use crate::time::Time;
use std::cmp::max;

const GRANULARITY_INV: usize = 10;

pub struct SingleSimulation {
    arrival_rate: usize,
    total_time: usize,
    router_queue: RouterQueue,
    packet_arrivals: HashMap<Time, usize>,
}

impl SingleSimulation {
    pub fn new(arrival_rate: usize) -> Self {
        let mut this = Self {
            arrival_rate,
            total_time: 1000,
            router_queue: RouterQueue::new(),
            packet_arrivals: Default::default(),
        };

        this.generate_times();
        this
    }

    fn reset(&mut self) {
        self.router_queue.reset();
        self.generate_times();
    }

    fn generate_times(&mut self) {
        self.packet_arrivals.clear();
        let num_packets = self.arrival_rate * self.total_time;
        let max_time = (self.total_time * GRANULARITY_INV) + 1;

        let rng = rand::thread_rng();
        let range = Uniform::from(0..max_time);

        rng.sample_iter(&range)
            .take(num_packets)
            .map(|t: usize| Time::init(t))
            .for_each(|t: Time| {
                match self.packet_arrivals.get(&t) {
                    Some(c) => self.packet_arrivals.insert(t, c + 1),
                    None => self.packet_arrivals.insert(t, 1)
                };
            });
    }

    pub fn simulate_and_reset(&mut self) -> (f64, usize) {
        let mut max_queue_size = 0;
        let max_time = (self.total_time * GRANULARITY_INV) + 1;

        let mut tick = 0;
        while tick < max_time || !self.router_queue.is_empty() {
            let t = Time::init(tick);
            let count = self.packet_arrivals.get(&t).cloned();
            for _ in 0..count.unwrap_or_default() {
                self.router_queue.push(Packet::new(&t));
            }

            max_queue_size = max(max_queue_size, self.router_queue.len());
            self.router_queue.pop(&t);
            tick += 1;
        }

        let res = (self.router_queue.avg_qd(), max_queue_size);
        self.reset();
        res
    }
}